src/services/base_service.js

```diff
@@ -0,0 +1,232 @@
+  plrep: 'Report',
+  prcl: 'Parcel',
+  rate: 'Rate',
+  refrep: 'Report',
+  rfnd: 'Refund',
+  sf: 'ScanForm',
+  shp: 'Shipment',
+  shpinvrep: 'Report',
+  shprep: 'Report',
+  trk: 'Tracker',
+  trkrep: 'Report',
+  user: 'User',
+};
+
+const RESOURCES = {
+  Address,
+  ApiKey,
+  Batch,
+  Brand,
+  CarrierAccount,
+  CarrierType,
+  CustomsInfo,
+  CustomsItem,
+  EasyPostObject,
+  EndShipper,
+  Event,
+  Form,
+  Insurance,
+  Order,
+  Parcel,
+  Payload,
+  Pickup,
+  PickupRate,
+  PostageLabel,
+  Rate,
+  Refund,
+  Report,
+  ScanForm,
+  Shipment,
+  Tracker,
+  User,
+  Webhook,
+};
+
+export default (easypostClient) =>
+  class BaseService {
+    static _url = null;
+
+    static _name = null;
+
+    static key = null;
+
+    /**
+     * Converts a response and all its nested elements to its associated EasyPostObject.
+     * @param {*} response
+     * @returns {*}
+     */
+    static convertToEasyPostObject(response) {
+      if (Array.isArray(response)) {
+        const mapped = [];
+        response.forEach((value, object) => {
+          if (typeof object === 'string' && RESOURCES[value] !== undefined) {
+            // eslint-disable-next-line no-param-reassign
+            value.object = object;
+          }
+          mapped.push(this.convertToEasyPostObject(value));
+        });
+
+        return mapped;
+        // eslint-disable-next-line no-else-return
+      } else if (typeof response === 'object' && response !== null) {
+        let className;
+        if (RESOURCES[response.object] !== undefined) {
+          className = RESOURCES[response.object];
+        } else if (
+          response.id !== undefined &&
+          EASYPOST_OBJECT_ID_PREFIX_TO_CLASS_NAME_MAP[
+            response.id.substr(0, response.id.indexOf('_'))
+          ] !== undefined
+        ) {
+          className =
+            EASYPOST_OBJECT_ID_PREFIX_TO_CLASS_NAME_MAP[
+              response.id.substr(0, response.id.indexOf('_'))
+            ];
+        } else {
+          className = 'EasyPostObject';
+        }
+        return this.constructFrom(response, className);
+      }
+      return response;
+    }
+
+    /**
+     * Construct an EasyPostObject from the various properties of a response.
+     * @param {*} values
+     * @param {*} className
+     * @returns
+     */
+    static constructFrom(values, className) {
+      const object = new RESOURCES[className.name !== undefined ? className.name : className]();
+      const convertedObject = this.mapProps(object, values);
+
+      return convertedObject;
+    }
+
+    /**
+     * Creates a new NotImplementedError.
+     * @param {string} functionName
+     * @returns {Promise<never>}
+     */
+    static notImplemented(functionName) {
+      return Promise.reject(new NotImplementedError(functionName, this._url));
+    }
+
+    /**
+     * Creates an EasyPost Object via the API.
+     * @param {*} params
+     * @returns {Base}
+     */
+    static async create(params) {
```

Comment:
There's something to be said for not having any "create", "all" or "retrieve" functions in the base class (that you have to override to disable in subclasses), and instead re-implementing these functions directly in the appropriate service classes.

Reply:
True - I did remove `delete` from the base class. This was the least amount of work required since this is how it's been. For services that don't allow one of create/all/retrieve, we add the `notImplemented` function to those.

Reply:
But is that how we want to continue doing it, having functions in here that are "not implemented"?

Reply:
For a user perspective, I think it's kind of ugly to have a function available, only for it to spit out "sorry, this actually doesn't do anything".

Reply:
I'm not necessarily opposed to changing it, if we do, we should handle it separately in a new PR?

Reply:
We can if you want, to avoid blocking this PR. I'll take it on.